File: backend/app/models/department.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class EmployeeView(BaseModel):
    employee_id: str
    name: str
    role: str
    monthly_salary: int
    manager_id: Optional[str]
    children_ids: list[str]
    direct_report_count: int
    team_headcount: int
    team_payroll: int


class DepartmentTotalsView(BaseModel):
    employee_count: int
    total_payroll: int


class RollupView(BaseModel):
    team_headcount: int
    team_payroll: int


class RollupChangeView(BaseModel):
    employee_id: str
    name: str
    role: str
    before: RollupView
    after: RollupView


class TransferImpactView(BaseModel):
    employee_id: str
    employee_name: str
    old_manager_id: str
    new_manager_id: str
    moved_subtree_ids: list[str]
    moved_headcount: int
    moved_payroll: int
    changed_rollup_ids: list[str]
    changes: list[RollupChangeView]
    root_unchanged: bool


class DepartmentView(BaseModel):
    scenario: str
    root_id: str
    employees: list[EmployeeView]
    totals: DepartmentTotalsView
    last_successful_transfer: Optional[TransferImpactView]


class ScenarioView(BaseModel):
    key: str
    label: str
    kind: str
    description: str


class TransferRequest(BaseModel):
    employee_id: str
    new_manager_id: str


class AddEmployeeRequest(BaseModel):
    employee_id: str
    name: str
    role: str
    monthly_salary: int
    manager_id: str


class LoadRequest(BaseModel):
    scenario: str = "main-12"

File: frontend/src/types/department.ts
export interface EmployeeView {
  employee_id: string
  name: string
  role: string
  monthly_salary: number
  manager_id: string | null
  children_ids: string[]
  direct_report_count: number
  team_headcount: number
  team_payroll: number
}

export interface DepartmentTotalsView {
  employee_count: number
  total_payroll: number
}

export interface RollupView {
  team_headcount: number
  team_payroll: number
}

export interface RollupChangeView {
  employee_id: string
  name: string
  role: string
  before: RollupView
  after: RollupView
}

export interface TransferImpactView {
  employee_id: string
  employee_name: string
  old_manager_id: string
  new_manager_id: string
  moved_subtree_ids: string[]
  moved_headcount: number
  moved_payroll: number
  changed_rollup_ids: string[]
  changes: RollupChangeView[]
  root_unchanged: boolean
}

export interface DepartmentView {
  scenario: string
  root_id: string
  employees: EmployeeView[]
  totals: DepartmentTotalsView
  last_successful_transfer: TransferImpactView | null
}

export interface ScenarioView {
  key: string
  label: string
  kind: string
  description: string
}

export interface TransferResponse {
  department: DepartmentView
  impact: TransferImpactView
}

export interface PreviewTransferResponse {
  impact: TransferImpactView
  department: DepartmentView
}

export interface AddEmployeeRequest {
  employee_id: string
  name: string
  role: string
  monthly_salary: number
  manager_id: string
}

export class ApiError extends Error {
  readonly code: string

  constructor(code: string, message: string) {
    super(message)
    this.name = 'ApiError'
    this.code = code
  }
}
